### ISSUE TITLE
deeplearning - 3/w2

### ISSUE THREAD

**Issue by getElementsByName** [getElementsByName/tmp-doc#3]
# Structuring Machine Learning Projects - Week 2. ML strategy (2)

## 주요 개념
- Understand what `multi-task learning` and `transfer learning` are
- Recognize bias, variance and `data-mismatch` by looking at the performances of your algorithm on train/dev/test sets

### Carrying out `error analysis`
- a set of `mislabeled examples`(`false` positive + `false` negative), either in your `dev` set, or in your development set.
- `hand engineering things`
   - I know that in machine learning, sometimes we speak disparagingly of hand engineering things, or using too much value insight. But if you're building applied systems, then this simple counting procedure, error analysis, can save you a lot of time.
- `the ceiling on performance`
    - `Upper bound` on how much you could `improve performance` by working on a specific problem
- This, in just 5 to 10 minutes, gives you an estimate of how `worthwhile` this direction is.

<details>
  <summary>예시
</summary>

![image](https://user-images.githubusercontent.com/21363169/86311831-3b87f380-bc5c-11ea-86f3-e43c3807ca83.png)

</details>

![image](https://user-images.githubusercontent.com/21363169/86311849-42af0180-bc5c-11ea-8dd2-70a24a682d38.png)

### Cleaning up incorrectly labeled data
- `mislabeled`, `incorrectly labeled`
   - So I've used the term, `mislabeled` examples, to refer to if your `learning algorithm outputs` the wrong value of Y. 
   - But I'm going to say, `incorrectly labeled` examples, to refer to if in the data set you have in the training set or the dev set or the test set, the label for Y, whatever a `human label assigned` to this piece of data, is actually incorrect.

#### training set
- robust to (near) `random` errors 
   - probably okay to just leave the errors
   - <-> systemic errors (consistently wrong)

#### test set
- make the dev/test sets come from `same distribution`
- examining examples your algorithm got `right` as well as ones it get wrong
- training / dev,test sets may come from slightly different distributions.
   - (later chpater) data mismatch -> training-dev set

<details>
  <summary>error analysis에 'incorrectly labeled' 고려하는 경우
</summary>

![image](https://user-images.githubusercontent.com/21363169/86311903-62462a00-bc5c-11ea-943c-aaa1fbe570dc.png)

</details>

### Build your first system quickly, then iterate
-  if you are tackling a new problem for the first time
   -  to figure out all the different directions you can go in, which ones are actually `the most worthwhile`
-   This advice applies less strongly if you're working on an application area in which you have significant prior experience.
   -  ex) if there's a significant body of academic literature

<details>
  <summary>
Guideline: Build your first system quickly, then iterate
</summary>

![image](https://user-images.githubusercontent.com/21363169/86312161-0f20a700-bc5d-11ea-8b63-391663095081.png)'

</details>

### Training and testing on different distributions
- (previous chapter) 
   - dev/test sets (+ validation metric): defining `what target` you want to aim at.
   - The way you choose your training set will affect `how well` you can actually hit that target
- a huge hunger for training data
- `training` on data that comes from a `different distribution` than your `dev and test sets`

![image](https://user-images.githubusercontent.com/21363169/86312349-7a6a7900-bc5d-11ea-8166-404540be2d18.png)

<details>
  <summary>
Speech recognition example
</summary>

![image](https://user-images.githubusercontent.com/21363169/86312388-8bb38580-bc5d-11ea-8eb6-8013da1c0b84.png)

</details>

### Bias and Variance with mismatched data distributions
- train-dev set, train-dev set error

![image](https://user-images.githubusercontent.com/21363169/86312451-b3a2e900-bc5d-11ea-9d42-24bc22fd12d6.png)
![image](https://user-images.githubusercontent.com/21363169/86312461-b998ca00-bc5d-11ea-8e89-47f8cc06687d.png)
![image](https://user-images.githubusercontent.com/21363169/86312465-bbfb2400-bc5d-11ea-970b-d794532ab02e.png)

### Addressing data mismatch
- artificial data synthesis
   - And artificial data synthesis does work. In speech recognition, I've seen artificial data synthesis significantly boost the performance of what were already very good speech recognition system. So, it can work very well. 
   - overfit to a small set: But, if you're using artificial data synthesis, just be cautious and bear in mind whether or not you might be accidentally simulating data only from a tiny subset of the space of all possible examples. 

#### quiz
<details>
  <summary>10)  If the synthesized images look realistic, then the model will just see them as if you had added useful data to identify road signs and traffic signals in a foggy weather. I will very likely help.
</summary>

![Screenshot from 2020-06-30 21-28-31](https://user-images.githubusercontent.com/21363169/86126079-adefbb00-bb18-11ea-8807-d0680a81f934.png)

</details>
![image](https://user-images.githubusercontent.com/21363169/86312512-d7fec580-bc5d-11ea-8df0-f759e6799e1f.png)
![image](https://user-images.githubusercontent.com/21363169/86312580-02e91980-bc5e-11ea-8513-281a68b0888e.png)

### Transfer learning
- pre-training 
   - pre-initialize
- fine tuning

![image](https://user-images.githubusercontent.com/21363169/86312674-3330b800-bc5e-11ea-94fb-aa5b9ba6a23b.png)
![image](https://user-images.githubusercontent.com/21363169/86312690-3a57c600-bc5e-11ea-8730-730bd77a9884.png)

#### 참고자료
- https://towardsdatascience.com/transfer-learning-from-pre-trained-models-f2393f124751

#### Feature Extraction -> Fine Tuning
- pretrained model에 부분 layer만 trainable 설정 + trainable 토글
   - https://keras.io/guides/transfer_learning/
   - https://www.tensorflow.org/tutorials/images/transfer_learning
- pretrained model `trainable = True` 설정 + 연속적 learning rate 튜닝
  - https://colab.research.google.com/notebooks/tpu.ipynb

### Multi-task learning
- In multi-task learning, you start off simultaneously, trying to have one neural network do several things at the same time. 
- And then each of these task helps hopefully all of the other task.
   - If some of the `earlier features` in neural network can `be shared` between these different types of objects, then you find that training one neural network to do four things results in better performance than training four completely separate neural networks to do the four tasks separately.
- you just `omit unlabel` term from summation but just sum over only the values where there is a label.
- the main difference between this and softmax regression, is that unlike softmax regression, which assigned a single label to single example. This one image can have multiple labels.

![image](https://user-images.githubusercontent.com/21363169/86312717-48a5e200-bc5e-11ea-8bf9-33a7c2df27c0.png)
![image](https://user-images.githubusercontent.com/21363169/86312734-565b6780-bc5e-11ea-83e0-8ed171f56c00.png)

### What is end-to-end deep learning?
- multi-step approach <-> end-to-end
   - multi-step approach
      - each of the two problems you're solving is actually much simpler.
      - a lot of data for each of the two sub-tasks.
- machine learning <-> hand-designed

![image](https://user-images.githubusercontent.com/21363169/86312951-e1d4f880-bc5e-11ea-8c00-d4e27c32167f.png)
![image](https://user-images.githubusercontent.com/21363169/86313000-ffa25d80-bc5e-11ea-8dc6-8725d63d52e7.png)
![image](https://user-images.githubusercontent.com/21363169/86313009-0630d500-bc5f-11ea-8191-f047554553ff.png)

### Whether to use end-to-end deep learning
- I think of a learning algorithm as having `two main sources` of knowledge. One is the `data` and the other is whatever you `hand design`, be components, or features, or other things.
- `hand-designed` system can actually allow humans to `inject a lot of knowledge` about the problem into an algorithm that should be very helpful.
   - hand-designed components tend to help more when you're training on a `small training set`
   - it's kind of `a double edged sword` that could hurt or help but it does tend to help more

![image](https://user-images.githubusercontent.com/21363169/86313135-5c9e1380-bc5f-11ea-99de-095f5d060824.png)
![image](https://user-images.githubusercontent.com/21363169/86313164-69bb0280-bc5f-11ea-9633-15a93da69ac1.png)

**Comment by getElementsByName** [comment]
## Transfer Learning
- https://towardsdatascience.com/transfer-learning-from-pre-trained-models-f2393f124751

### Feature Extraction -> Fine Tuning
- pretrained model 전체 trainable 토글
   - https://keras.io/guides/transfer_learning/
- pretrained model에 부분 layer만 trainable 설정 + trainable 토글
   - https://www.tensorflow.org/tutorials/images/transfer_learning
- pretrained model `trainable = True` 설정 + 연속적 learning rate 튜닝
  - https://colab.research.google.com/notebooks/tpu.ipynb

**Comment by getElementsByName** [comment]
### quiz
<details>
  <summary>10)  If the synthesized images look realistic, then the model will just see them as if you had added useful data to identify road signs and traffic signals in a foggy weather. I will very likely help.
</summary>

![Screenshot from 2020-06-30 21-28-31](https://user-images.githubusercontent.com/21363169/86126079-adefbb00-bb18-11ea-8807-d0680a81f934.png)

</details>

**Comment by getElementsByName** [comment]
- previous chapter 
   - dev/test sets (+ validation metric): defining `what target` you want to aim at.
   - The way you choose your training set will affect `how well` you can actually hit that target
